### ISSUE TITLE
[v2] StaticQuery doesn't work with typescript

### ISSUE THREAD

**Issue by dannywils** [gatsbyjs/gatsby#5626]
## Description

StaticQuery's are forever stuck in the "Loading (StaticQuery)" state in both development and production build when using typescript. 

### Steps to reproduce

I've created a minimal example reproducing the issue. 

https://github.com/dannywils/gatsby-typescript-staticquery

### Expected result

![image](https://user-images.githubusercontent.com/3779698/40783317-6b31e15c-64b8-11e8-8e63-cf9aeeb65162.png)
The StaticQuery should be extracted, executed and the render function should be called with the query results. 

### Actual result

What happened.

### Environment
* Gatsby version: 2.0.0-alpha.43
* gatsby-cli version: 1.1.53
* Node.js version: v10.0.0
* Operating System: Windows 10 (16299.431)

### File contents (if changed)
`gatsby-config.js`: https://github.com/dannywils/gatsby-typescript-staticquery/blob/master/gatsby-config.js
`package.json`: https://github.com/dannywils/gatsby-typescript-staticquery/blob/master/package.json
`gatsby-node.js`: https://github.com/dannywils/gatsby-typescript-staticquery/blob/master/gatsby-node.js
`gatsby-browser.js`: N/A
`gatsby-ssr.js`: N/A

**Comment by dannywils** [comment]
I've narrowed this down to not being a typescript issue, but actually due to JSX being transpiled before it hits the file parser. 

Using static queries without JSX example does not work because the resulting AST is different than expected here: https://github.com/gatsbyjs/gatsby/blob/v2/packages/gatsby/src/internal-plugins/query-runner/file-parser.js#L119

``
     {React.createElement(StaticQuery, {
        query: graphql`
          query LayoutQuery {
            site {
              siteMetadata {
                title
              }
            }
          }
        `,
        render: data => data.site.siteMetadata.title
      })}
```

Should the file parser be modified to extract graphql queries from createElement calls?

**Comment by KyleAMathews** [comment]
>Should the file parser be modified to extract graphql queries from createElement calls?

Yeah — that seems like a good solution. Would help with other compile-to-js languages as well.

**Comment by KyleAMathews** [comment]
Want to do a PR?

**Comment by mquandalle** [comment]
Wouldn't the switch from `ts-loader` to `@babel/preset-typescript` discussed in #5520 fix this issue?

**Comment by dannywils** [comment]
Yeah that should do it! For ts-loader setting JSX to "preserve" in compilerOptions fixes the issue as well because babel transpiles the JSX after the static queries are extracted.

https://www.typescriptlang.org/docs/handbook/jsx.html

If the short-term plan is to switch to babel's typescript preset then no further action should be required.

**Comment by ascorbic** [comment]
I'm still getting this with beta.11. I get Loading (StaticQuery) in development, and in production I get 

`Error: ./src/pages/index.tsx
  Module not found: Error: Can't resolve '../../public/static/d/2999245696.json' in '/Users/matt/Documents/repos/site-rebuild/src/pages'`

I've changed jsx to preserve, with no effect.

**Comment by stasgavrylov** [comment]
I just wanted to point out that this bug is still reproducible.
I'm using `gatsby@2.0.46` and `gatsby-plugin-typescript@2.0.1`.

StaticQuery in TS component works fine in dev mode, but produces `Loading (StaticQuery)` in build output.

I wish I'd find this issue earlier. Changing my component to plain JS fixed the problem.

**Comment by portenez** [comment]
Hi, this issue is still happening on `gatsby@2.18.4` + `gatsby-plugin-typescript@2.1.20`. I guess we probably need to open a new ticket.
